File: solutions/typescript/binary-search-tree/1/binary-search-tree.ts
class Node {
  constructor(
    public data: number,
    public left: Node | null = null,
    public right: Node | null = null) { }

}
export class BinarySearchTree {
  private root: Node

  constructor(data: number) {
    this.root = new Node(data)
  }

  public get data(): number {
    return this.root.data
  }

  public get right(): Node | null {
    return this.root.right
  }

  public get left(): Node | null {
    return this.root.left
  }

  public insert(item: number, root = this.root): void {
    if (item <= root.data) {
      if (root.left) {
        this.insert(item, root.left)
      } else {
        root.left = new Node(item)
      }
    } else if (item > root.data) {
      if (root.right) {
        this.insert(item, root.right)
      } else {
        root.right = new Node(item)
      }
    }
  }

  public each(callback: (data: number) => void): void {
    this.inOrderTraversal(this.root).forEach(callback)
  }

  private inOrderTraversal(root: Node | null = this.root): number[] {
    if (!root) {
      return []
    }

    return [
      ...this.inOrderTraversal(root.left),
      root.data,
      ...this.inOrderTraversal(root.right),
    ]
  }
}

File: solutions/c/binary-search-tree/1/binary_search_tree.c
#include "binary_search_tree.h"
#include <stdlib.h>

static node_t *new_node(int data) {
    node_t *node = (node_t *)malloc(sizeof(node_t));
    node->data = data;
    node->left = NULL;
    node->right = NULL;
    return node;
}

static size_t count_nodes(node_t *node) {
    if (node == NULL) {
        return 0;
    }
    return 1 + count_nodes(node->left) + count_nodes(node->right);
}

static void insert_node(node_t *node, int data) {
    if (data <= node->data) {
        if (node->left == NULL) {
            node->left = new_node(data);
        } else {
            insert_node(node->left, data);
        }
    }
    else {
        if (node->right == NULL) {
            node->right = new_node(data);
        } else {
            insert_node(node->right, data);
        }
    }
}

static void inorder_traversal(node_t *node, int *sorteddata, size_t *size) {
    if (node == NULL) {
        return;
    }
    inorder_traversal(node->left, sorteddata, size);
    sorteddata[*size] = node->data;
    (*size)++;
    inorder_traversal(node->right, sorteddata, size);
}


node_t *build_tree(int *tree_data, size_t tree_data_len) {
    node_t *root = new_node(tree_data[0]);
    for (size_t i = 1; i < tree_data_len; i++) {
        insert_node(root, tree_data[i]);
    }
    return root;
}



void free_tree(node_t *tree) {
    if (tree == NULL) {
        return;
    }
    free_tree(tree->left);
    free_tree(tree->right);
    free(tree);
}

int *sorted_data(node_t *tree) {
    if (tree == NULL) {
        return NULL;
    }
    size_t size = 0;
    size_t index = 0;

    size = count_nodes(tree);
    int *sorteddata = (int *)malloc(sizeof(int) * size);
    inorder_traversal(tree, sorteddata, &index);
    return sorteddata;
}
